### PR TITLE
[GH-20818] - Add restore run feature for playbook

### PR DESCRIPTION
**Pull request by rolwin100**
#### Summary

* This feature restores a run that is finished.
* There was a /restore feat already present in the repo
* created a api request in client.ts
* added a hook in restore_run.tsx
* Added menu in button in context_menu.tsx and called the hook on button click.

#### Ticket Link

https://github.com/mattermost/mattermost-server/issues/20818

#### Checklist

- [ ] ~~Telemetry updated~~
- [ ] ~~Gated by experimental feature flag~~
- [ ] ~~Unit tests updated~~

### CODE REVIEW

File: webapp/i18n/en.json [hunk 1]
```diff
@@ -144,6 +144,7 @@
   "DtCplA": "{numParticipants, plural, =1 {<b>#</b> participant} other {<b>#</b> participants}}",
   "E0LnBo": "You can select an option or specify a custom duration (\"2 weeks\", \"3 days 12 hours\", \"45 minutes\", ...)",
   "EC5MJD": "There are no updates available.",
+  "ELjMDy": "Are you sure you want to restart the run ?",
```
Comment (nickmisasi) [inline webapp/i18n/en.json:147]:
```suggestion
  "ELjMDy": "Are you sure you want to restart the run?",
```

File: webapp/src/components/backstage/playbook_runs/playbook_run/restore_run.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import {makeUncontrolledConfirmModalDefinition} from 'src/components/widgets/confirmation_modal';
+import {useLHSRefresh} from '../../lhs_navigation';
+
+export const useOnRestoreRun = (playbookRun: PlaybookRun) => {
+    const dispatch = useDispatch();
+    const {formatMessage} = useIntl();
+    const refreshLHS = useLHSRefresh();
+
+    return () => {
+        const confirmationMessage = formatMessage({defaultMessage: 'Are you sure you want to restart the run ?'});
```
Comment (nickmisasi) [inline webapp/src/components/backstage/playbook_runs/playbook_run/restore_run.tsx:19]:
```suggestion
        const confirmationMessage = formatMessage({defaultMessage: 'Are you sure you want to restart the run?'});
```

**Review by nickmisasi — CHANGES_REQUESTED** [review]
Overall LGTM! Couple changes required.

My question here is whether this is _restoring_ or _restarting_ the run. Will the run pick up where it was left off (ie, any already completed steps will still be completed) or will this act more like it's a fresh new run?

Depending on which this is, I think it's important to ensure that all verbiage (things displayed to user, as well as file names/function names/etc) is consistently either `restore` _or_ `restart` to avoid confusion.

**Comment by itao** [comment]
👋 Hi @rolwin100, thanks for taking on this PR! I noticed that the [original issue](https://github.com/mattermost/mattermost-server/issues/20818) mentioned

> Restarting a run should add an event to the timeline, as well as post a message to the associated channel:

Could you confirmed if that was something already implemented or if it's expected as part of this PR, please?

**Comment by rolwin100** [comment]
Its'a part of this PR.

File: webapp/i18n/en.json (hunk 1 above)
Comment (rolwin100) [reply webapp/i18n/en.json:147]: @nickmisasi this feature will restore as well as restart the run as per what I can understand.

Comment (rolwin100) [reply webapp/i18n/en.json]: Made a commit and removed the extra space.

File: webapp/src/components/backstage/playbook_runs/playbook_run/restore_run.tsx (hunk 2 above)
Comment (rolwin100) [reply webapp/src/components/backstage/playbook_runs/playbook_run/restore_run.tsx]: Made a commit and removed the extra space.

**Comment by itao** [comment]
@nickmisasi regarding your question whether this is restoring or restarting the run.

Yes, the run will pick up where it was left off. However I prefer the copy `restart` over `restore` since the latter is more often associated with archiving or deletion, whereas the action we're undoing in this case is finishing. There is still a difference between:
- `restart the run`: pick up where it left off when marked as finished
- `rerun the playbook`: start a fresh new run

**Comment by rolwin100** [comment]
> @nickmisasi regarding your question whether this is restoring or restarting the run.
> 
> Yes, the run will pick up where it was left off. However I prefer the copy `restart` over `restore` since the latter is more often associated with archiving or deletion, whereas the action we're undoing in this case is finishing. There is still a difference between:
> 
> * `restart the run`: pick up where it left off when marked as finished
> * `rerun the playbook`: start a fresh new run

@itao @nickmisasi  can we proceed with the existing change because the api which was present was named as restore and I don't see much problem with this.

File: webapp/src/components/backstage/playbook_runs/playbook_run/restore_run.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            show: true,
+            title: formatMessage({defaultMessage: 'Confirm restart run'}),
+            message: confirmationMessage,
+            confirmButtonText: formatMessage({defaultMessage: 'Restart run'}),
+            onConfirm,
+            // eslint-disable-next-line @typescript-eslint/no-empty-function
+            onCancel: () => {},
+        })));
+    };
+};
```
Comment (gbochora) [inline webapp/src/components/backstage/playbook_runs/playbook_run/restore_run.tsx:36]: New line at the end of the file.
```suggestion
};

```

File: webapp/src/components/backstage/playbook_runs/playbook_run/restore_run.tsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        };
+
+        dispatch(modals.openModal(makeUncontrolledConfirmModalDefinition({
+            show: true,
+            title: formatMessage({defaultMessage: 'Confirm restart run'}),
+            message: confirmationMessage,
+            confirmButtonText: formatMessage({defaultMessage: 'Restart run'}),
+            onConfirm,
+            // eslint-disable-next-line @typescript-eslint/no-empty-function
+            onCancel: () => {},
```
Comment (gbochora) [inline webapp/src/components/backstage/playbook_runs/playbook_run/restore_run.tsx:33]:
```suggestion
            onCancel: () => null,
```

**Review by gbochora — APPROVED** [review]
Great work, @rolwin100! Thanks! I left two optional suggestions.

File: webapp/src/components/backstage/playbook_runs/playbook_run/restore_run.tsx (hunk 4 above)
Comment (rolwin100) [reply webapp/src/components/backstage/playbook_runs/playbook_run/restore_run.tsx]: Yo @gbochora thanks for the suggestion :). Yes, this was a must.

**Comment by gbochora** [comment]
@rolwin100, great work on tests; thanks again! We will be able to merge this PR after the second reviewer approves it.

File: tests-e2e/cypress/integration/runs/rdp_main_restore_spec.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        }).then((playbookRun) => {
+            testRun = playbookRun;
+
+            // # Visit the playbook run
+            cy.visit(`/playbooks/runs/${testRun.id}`);
+        });
+    });
+
+    describe('restart run', () => {
+        it('can be confirmed', () => {
```
Comment (nickmisasi) [inline tests-e2e/cypress/integration/runs/rdp_main_restore_spec.js:80]: This test seems to be failing: 
image

**Review by nickmisasi — COMMENTED** [review]
Code LGTM. Will need tests to pass before approval. Thanks @rolwin100 🎉!

File: tests-e2e/cypress/integration/runs/rdp_main_restore_spec.js (hunk 5 above)
Comment (rolwin100) [reply tests-e2e/cypress/integration/runs/rdp_main_restore_spec.js:80]: let me check it again it's working for me let me cross check and verify it

Comment (rolwin100) [reply tests-e2e/cypress/integration/runs/rdp_main_restore_spec.js:80]: I'm not facing this issue. I'm running  `CYPRESS_developerMode=true npm run cypress:open` inside playbook tests-e2e folder as mentioned by @trilopin

Comment (Phrynobatrachus) [reply tests-e2e/cypress/integration/runs/rdp_main_restore_spec.js:80]: Fails about 20% of the time locally for me. In the test, the run is finished and _then_  you check for the `In Progress` badge, this ordering makes the assertion dependent on how quickly the API resolves. If it takes a bit longer, the text will still be `In Progress` and the assertion works, if it completes quickly the badge will have already changed to `Finished` and the assertion fails.   

Move the `In Progress` assertion above the finish run modal interactions and it should be fine. :+1:

Comment (rolwin100) [reply tests-e2e/cypress/integration/runs/rdp_main_restore_spec.js:80]: Oops Got it my bad :) thanks @Phrynobatrachus

Comment (rolwin100) [reply tests-e2e/cypress/integration/runs/rdp_main_restore_spec.js:80]: @nickmisasi this is resolved

**Review by nickmisasi — APPROVED** [review]
Thanks @rolwin100!

**Review by Phrynobatrachus — APPROVED** [review]

**Pull request merged**